### DISCUSSION TITLE
是否可通过某种方法只使用 SNI 伪造功能？

### DISCUSSION THREAD

**Discussion by 185264646** [SeaHOH/GotoX#167] [问答/咨询]
这个项目也好久没更新了，我也不打算用这里面的GAE和CFW啥的。只想用这玩意绕过SNI阻断，不挂梯子连接avatars3.githubusercontent.com。有办法进行配置吗？

**Comment by SeaHOH** [comment]
> 只想用这玩意绕过SNI阻断，不挂梯子连接avatars3.githubusercontent.com。有办法进行配置吗？

单只用 SNI 伪造可以这样配置：
1. 关闭默认开启的 GWS IP 筛选
1. 配置 [filter] 默认动作为 2 (forward)
1. 清空你不需要的默认自动规则，注意 DoH 相关不要删除
1. 配置你的 SNI 伪造规则
1. 可能的话建议通过浏览器扩展智能规则调用，这样其它无需伪造的连接可减少一次转发

这里有个问题，`.githubusercontent.com` 似乎并未被阻断，我这里是能够直连的。

**Reply by 185264646** [reply]
已经试了一下，尽管没太搞懂，大概已经成功把对avatars3.githubusercontent.com的TLS连接的SNI头直接去掉了，但是仍然无法连接，应该是github那边的SNI检测策略较为严格的原因。
GotoX这边报错：
![图片](https://user-images.githubusercontent.com/43387518/103002457-ef15d480-4569-11eb-9c53-105db28a8bbd.png)

**Reply by SeaHOH** [reply]
你这是浏览器连接代理后，加密连接验证失败。检查主配置中的 localssl 是否与浏览器设置相符，一般来讲留空使用默认值就行。

**Reply by 185264646** [reply]
> 
> 
> 你这是浏览器连接代理后，加密连接验证失败。检查主配置中的 localssl 是否与浏览器设置相符，一般来讲留空使用默认值就行。

GotoX的Config.ini的localssl项保持了留空，应该使用的是TLS1.2。

刚刚抓了下包，发现GotoX确实成功删除了Client Hello中的SNI信息，但是仍然被未知力量发送了RST，目前不清楚是github发的还是某神秘设备发的。
如果有必要的话我可以提供抓包结果以供分析。

**Reply by SeaHOH** [reply]
> 刚刚抓了下包，发现GotoX确实成功删除了Client Hello中的SNI信息，但是仍然被未知力量发送了RST，目前不清楚是github发的还是某神秘设备发的。

确认下，你这段描述和上面的贴图不是同样的错误吧。因为前一个错误就像我之前解释的，在根本还没对外发出连接，连接 GotoX 时就已经失败了。

如果确实是连接 `.githubusercontent.com` 时仍然报错，请查看连接中记录的 IP，或切换到 log level 2 查看 DNS 解析记录， 确认 IP 是正确的，大概有以下这些。可尝试优先使用 DoH 解析。
```
151.101.0.133
151.101.16.133
151.101.28.133
151.101.36.133
151.101.64.133
151.101.108.133
151.101.128.133
151.101.192.133
151.101.228.133
```

**Reply by 185264646** [reply]
> 
> 
> > 刚刚抓了下包，发现GotoX确实成功删除了Client Hello中的SNI信息，但是仍然被未知力量发送了RST，目前不清楚是github发的还是某神秘设备发的。
> 
> 确认下，你这段描述和上面的贴图不是同样的错误吧。因为前一个错误就像我之前解释的，在根本还没对外发出连接，连接 GotoX 时就已经失败了。
> 
> 如果确实是连接 `.githubusercontent.com` 时仍然报错，请切换到 log level 2， 确认连接的 IP 是正确的，大概有以下这些。可尝试优先使用 DoH 解析。
> 
> ```
> 151.101.0.133
> 151.101.16.133
> 151.101.28.133
> 151.101.36.133
> 151.101.64.133
> 151.101.108.133
> 151.101.128.133
> 151.101.192.133
> 151.101.228.133
> ```

DNS查询结果没有问题，我的默认DNS服务器没有对结果进行污染，查询结果151.101.76.133
可能是我的ActionFilter.ini配置有问题。貌似它在不停地嵌套TLS查询，发现很多TLS1的Client Hello里面又套了一个TLS1.2的Client Hello被发送到151.101.76.133。

**Reply by SeaHOH** [reply]
Client Hello 是由底层构造并发送，GotoX 只是设置不发送 SNI，或发送伪造的名称，有可能会重试握手验证，但不会嵌套 (查询？？)。你配置的应该是最简单的不发送吧。
```ini
[0-fakecert]
.githubusercontent.com = @none
```

**Reply by 185264646** [reply]
确实是不发送SNI信息，刚刚是我搞错了，只是外层的Record Layer使用了TLS1.0而已，并不是嵌套。我贴张图吧：
![图片](https://user-images.githubusercontent.com/43387518/103092222-842fd080-4631-11eb-86e3-9db99704fcb4.png)
你看Client Hello信息中不含SNI，但是下一个包就是RST了。而且这个RST包后面竟然还有Server Hello，只不过没啥意义了。

**Reply by SeaHOH** [reply]
看起来确实是在没有 SNI 的情况下被 RST 了，可能是针对 IP 的。

我确认下情况，你之前直连 `.githubusercontent.com` 是一直被重置吗，还是说有时候能连接上？

**Reply by SeaHOH** [reply]
我刚逐个测试了一下这些 IP，可连接、可连接但丢包严重、被重置，各种情况都有，你可以测试一遍选出可用的来配置。
```ini
[序号-direct]
.githubusercontent.com = IP1 | IP2 | ...
```

**Reply by notfound-hust** [reply]
> 看起来确实是在没有 SNI 的情况下被 RST 了，可能是针对 IP 的。
> 
> 我确认下情况，你之前直连 `.githubusercontent.com` 是一直被重置吗，还是说有时候能连接上？

avatars*.githubusercontent.com在极少数情况下能连上，原因未知，有时候一张图刷新五六遍能刷新出来。可能是神秘设备没反应过来(笑)。可能是移动自己架的，因为以前用联通就没有问题。不过似乎是针对ip和TLS的，因为raw.githubusercontent.com的资源非常容易下载。而且ping avatars3.githubusercontent.com效果非常好，几乎不丢包。直接访问80端口也每次都会被重定向到https再被RST，从没有在访问80端口的时候被阻断。
可能直接忽略这个RST会有效果，因为似乎对面没有收到RST包，后续的Server Hello也还能收到。不知道GotoX能不能做到这件事。

**Reply by SeaHOH** [reply]
**请不要换号回复，我有可能会分辨不出。**

---

所有 `.githubusercontent.com` 解析结果都是相同的，同一个 IP 不存在连通性的差别。
既然之前就能连接，那么逐个测试，选出效果较好的几个 IP 来使用。测试和配置都可以使用 direct 规则，无需伪造 SNI。

> 可能直接忽略这个RST会有效果，因为似乎对面没有收到RST包，后续的Server Hello也还能收到。不知道GotoX能不能做到这件事。

你可以试试 [TCPioneer](https://github.com/Macronut/TCPioneer) 或这个作者的另一个全平台工具，不过这种工具可能会失效比较快，需及时更新，或根据抓包结果重新分析配置 (默认配置不起作用时也是如此处理)。
